diff -urN a/package/feeds/packages/openvswitch/Makefile b/package/feeds/packages/openvswitch/Makefile
--- a/package/feeds/packages/openvswitch/Makefile	2020-12-15 14:05:57.848483999 +0530
+++ b/package/feeds/packages/openvswitch/Makefile	2020-12-15 14:05:57.848483999 +0530
@@ -222,7 +222,9 @@
 define ovs_common_install
 	$$(INSTALL_DIR) $$(1)/etc/openvswitch
 	$$(INSTALL_DIR) $$(1)/etc/init.d
+ifndef CONFIG_PACKAGE_opensync
 	$$(INSTALL_BIN) ./files/openvswitch.init $$(1)/etc/init.d/openvswitch
+endif
 	$$(INSTALL_DIR) $$(1)/etc/config
 	$$(INSTALL_DATA) ./files/openvswitch.config $$(1)/etc/config/openvswitch
 	$$(INSTALL_DIR) $$(1)/usr/share/openvswitch/scripts
diff -urN a/package/network/config/firewall/Makefile b/package/network/config/firewall/Makefile
--- a/package/network/config/firewall/Makefile	2020-12-15 14:05:57.844484001 +0530
+++ b/package/network/config/firewall/Makefile	2020-12-15 14:05:57.848483999 +0530
@@ -48,7 +48,9 @@
 	$(INSTALL_DIR) $(1)/sbin
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/firewall3 $(1)/sbin/fw3
 	$(INSTALL_DIR) $(1)/etc/init.d
+ifndef CONFIG_PACKAGE_opensync
 	$(INSTALL_BIN) ./files/firewall.init $(1)/etc/init.d/firewall
+endif
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/iface
 	$(INSTALL_CONF) ./files/firewall.hotplug $(1)/etc/hotplug.d/iface/20-firewall
 	$(INSTALL_DIR) $(1)/etc/config/
diff -urN a/package/network/services/dnsmasq/Makefile b/package/network/services/dnsmasq/Makefile
--- a/package/network/services/dnsmasq/Makefile	2020-12-15 14:05:57.848483999 +0530
+++ b/package/network/services/dnsmasq/Makefile	2020-12-15 14:05:57.848483999 +0530
@@ -165,7 +165,9 @@
 	$(INSTALL_CONF) ./files/dhcp.conf $(1)/etc/config/dhcp
 	$(INSTALL_CONF) ./files/dnsmasq.conf $(1)/etc/dnsmasq.conf
 	$(INSTALL_DIR) $(1)/etc/init.d
+ifndef CONFIG_PACKAGE_opensync
 	$(INSTALL_BIN) ./files/dnsmasq.init $(1)/etc/init.d/dnsmasq
+endif
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/dhcp
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/neigh
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/ntp
diff -urN a/package/network/utils/iproute2/Makefile b/package/network/utils/iproute2/Makefile
--- a/package/network/utils/iproute2/Makefile	2020-12-15 14:05:57.844484001 +0530
+++ b/package/network/utils/iproute2/Makefile	2020-12-16 12:57:27.867861946 +0530
@@ -155,6 +155,10 @@
 define Package/ip-full/install
 	$(INSTALL_DIR) $(1)/usr/libexec
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/ip/ip $(1)/usr/libexec/ip-full
+ifdef CONFIG_PACKAGE_opensync
+	$(INSTALL_DIR) $(1)/usr/sbin/
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/ip/ip $(1)/usr/sbin/ip
+endif
 endef
 
 define Package/tc/install
diff -urN a/package/feeds/packages/coreutils/Makefile b/package/feeds/packages/coreutils/Makefile
--- a/package/feeds/packages/coreutils/Makefile	2021-04-09 12:27:14.303376781 +0800
+++ b/package/feeds/packages/coreutils/Makefile	2021-04-09 12:27:57.031506251 +0800
@@ -65,11 +65,11 @@
 DEPENDS_expr = +libgmp
 DEPENDS_factor = +libgmp
 DEPENDS_cp = +libacl
-DEPENDS_dir = +libacl +libcap
+DEPENDS_dir = +libacl +libcap-ng
 DEPENDS_install = +libacl
-DEPENDS_ls = +libacl +libcap
+DEPENDS_ls = +libacl +libcap-ng
 DEPENDS_mv = +libacl
-DEPENDS_vdir = +libacl +libcap
+DEPENDS_vdir = +libacl +libcap-ng
 
 FILES_stdbuf := usr/lib/coreutils/libstdbuf.so
 
@@ -124,7 +124,7 @@
 	--disable-assert \
 	--disable-rpath \
 	--disable-libsmack \
-	--enable-libcap \
+	--enable-libcap-ng \
 	--without-linux-crypto \
 	--without-openssl \
 	--$(if $(CONFIG_USE_MUSL),with,without)-included-regex \
