--- a/src/lib/inet/src/linux/inet_gretap.c
+++ b/src/lib/inet/src/linux/inet_gretap.c
@@ -49,6 +49,17 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBI
 #define CONFIG_INET_GRE_USE_GRETAP
 #endif
 
+#ifdef CONFIG_OVS_SOFTWARE_ACCELERATION
+/*
+ * $1 - command (add|del)
+ * $2 - interface name
+ */
+static char nftool[] =
+_S(
+    nftool.sh -c "$1" -i "$2"
+);
+#endif
+
 /*
  * $1 - interface name
  * $2 - parent interface name
@@ -235,6 +246,15 @@ bool inet_gretap_interface_start(inet_gr
             return false;
         }
 
+#ifdef CONFIG_OVS_SOFTWARE_ACCELERATION
+        status = execsh_log(LOG_SEVERITY_INFO, nftool, "add", self->inet.in_ifname);
+        if (!WIFEXITED(status) || WEXITSTATUS(status) != 0)
+        {
+            LOG(ERR, "inet_gretap: %s: Error adding GRETAP interface to fastpath.", self->inet.in_ifname);
+            return false;
+        }
+#endif
+
         LOG(INFO, "inet_gretap: %s: GRETAP interface was successfully created.", self->inet.in_ifname);
     }
     else
@@ -245,6 +265,15 @@ bool inet_gretap_interface_start(inet_gr
             LOG(ERR, "inet_gretap: %s: Error deleting GRETAP interface.", self->inet.in_ifname);
         }
 
+#ifdef CONFIG_OVS_SOFTWARE_ACCELERATION
+        status = execsh_log(LOG_SEVERITY_INFO, nftool, "del", self->inet.in_ifname);
+        if (!WIFEXITED(status) || WEXITSTATUS(status) != 0)
+        {
+            LOG(ERR, "inet_gretap: %s: Error removing GRETAP interface from fastpath.", self->inet.in_ifname);
+            return false;
+        }
+#endif
+
         LOG(INFO, "inet_gretap: %s: GRETAP interface was successfully deleted.", self->inet.in_ifname);
     }
 
--- a/platform/cfg80211/src/lib/target/target_cfg80211.c
+++ b/platform/cfg80211/src/lib/target/target_cfg80211.c
@@ -2971,6 +2971,17 @@ util_vif_config_athnewind(const char *ph
 /******************************************************************************
  * Vif implementation
  *****************************************************************************/
+#ifdef CONFIG_OVS_SOFTWARE_ACCELERATION
+#include "execsh.h"
+/*
+ * $1 - command (add|del)
+ * $2 - interface name
+ */
+static char nftool[] =
+_S(
+    nftool.sh -c "$1" -i "$2"
+);
+#endif
 
 bool
 target_vif_config_set2(const struct schema_Wifi_VIF_Config *vconf,
@@ -3005,6 +3016,9 @@ target_vif_config_set2(const struct sche
             nl_req_init_channels(&target_nl_global, vif, g_chan_status);
             nl_req_del_iface(&target_nl_global, vif);
             util_vif_config_athnewind(phy);
+#ifdef CONFIG_OVS_SOFTWARE_ACCELERATION
+            execsh_log(LOG_SEVERITY_INFO, nftool, "del", vif);
+#endif
         }
 
         if (!rconf || !vconf->enabled)
@@ -3023,6 +3037,9 @@ target_vif_config_set2(const struct sche
              rconf->channel_exists ? rconf->channel : 0);
 
         nl_req_add_iface(&target_nl_global, vif, phy, vconf->mode, macaddr);
+#ifdef CONFIG_OVS_SOFTWARE_ACCELERATION
+        execsh_log(LOG_SEVERITY_INFO, nftool, "add", vif);
+#endif
 
         hostap_ctrl_discover(vif);
 
